src/security/Session.h

```diff
@@ -43,11 +44,15 @@ bool CreateServerSession(const Security::ContextPointer &, const Comm::Connectio
 #if USE_OPENSSL
 typedef SSL Connection;
 
+using Session = SSL_SESSION;
```

Comment:
Adding a generic abstraction here needs to be matched with the equivalent for all other supported libraries. It looks rather odd to have USE_OPENSSL wrappers around code elsewhere that uses only the abstractions, which use of a library-specific re-typename like this does.

Reply:
Yes the CommunicationSecrets members are a good example of such code. The .cc wrapping is okay as-is for now, but when the second library is added will become internal to the methods implemented. The .h lines shold not need or have wrapping.

Is there a reason to care about private members being declared but undefined when they are unused? AFAIK they should just be dropped by the compiler.

Reply:
> when the second library is added will become internal to the methods implemented

Yes, provided the second library will benefit from having the same private method(s).

> Is there a reason to care about private members being declared but undefined when they are unused?

I think there is. Without ifdefs, the reader and various automation tools may search for the method definition in vain. It is _usually_ not a big deal, but adding those ifdefs is not a big deal either. Lack of ifdefs will make certain coding bugs more difficult to find during development. And if all other factors were equal, I think being consistent across declarations and definitions is better than the alternative, especially when that consistency costs virtually nothing.

Reply:
IMO the reader / dev searching should find the ones inside ifdef and see what is going on. Automation we only need to care about if it has problems or complains. Otherwise it is a matter of style and clean looking code - which is where removing the ifdef from .h helps improve the look of code.

I do not insist on this change.

Reply:
> IMO the reader / dev searching should find the ones inside ifdef and see what is going on.

I see two wrinkles in this logic:

1. Just seeing the ifdefed code in the .cc file would not clarify anything for some. For me, for example, it would only raise questions like "Why were these method declarations left without ifdef protections? Is there another definition of these methods that I have missed?"
2. Modern developers often use automation to quickly "find" things -- they prefer not to grep for names but ask their IDE-like environment where the method is defined. That automation is getting pretty smart these days -- using compiler-produced namespace-aware maps for lookup. Some automation is aware of C++ macros. Unusual patterns may break automation (as discussed elsewhere), and such breakage may affect what developers can easily "find" or "see".

> Automation we only need to care about if it has problems or complains

Many automation failures are silent. For example, we do not know about all the cases where the compiler failed to optimize something away. Thus, we should not misinterpret absence of known problems as absence of problems.

> Otherwise it is a matter of style and clean looking code - which is where removing the ifdef from .h helps improve the look of code.

I agree that there is a "matter of style" component here as well! Unfortunately, we disagree which style is overall better. With all the other factors being equal (and they are not), to me, consistency and lack of surprises/exceptions/explanations is much better than a private class declaration section without ifdefs.

> I do not insist on this change.

Thank you.